plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/githubPullRequest.ts

```diff
@@ -0,0 +1,231 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs';
+import path from 'path';
+import { promisify } from 'util';
+
+import {
+  GithubCredentialsProvider,
+  ScmIntegrationRegistry,
+} from '@backstage/integration';
+import { zipObject } from 'lodash';
+import { createTemplateAction } from '../../createTemplateAction';
+import { Octokit } from '@octokit/rest';
+import { OctokitResponse } from '@octokit/types';
+import { InputError } from '@backstage/errors';
+import { createPullRequest } from 'octokit-plugin-create-pull-request';
+import globby from 'globby';
+
+const readFile = promisify(fs.readFile);
+
+type PullRequestResponse = {
+  html_url: string;
+};
+
+export interface PullRequestCreator {
+  createPullRequest(
+    options: createPullRequest.Options,
+  ): Promise<OctokitResponse<PullRequestResponse> | null>;
+}
+
+export type GithubPullRequestActionInput = {
+  owner: string;
+  repo: string;
+  title: string;
+  branchName: string;
+  description: string;
+  host?: string;
+  targetPath?: string;
+  sourcePath?: string;
+};
+
+export type ClientFactoryInput = {
+  host: string;
+  owner: string;
+  repo: string;
+};
+
+export const createClientFactory = async (
+  integrations: ScmIntegrationRegistry,
+  { owner, repo, host = 'github.com' }: ClientFactoryInput,
+): Promise<Octokit> => {
+  const integrationConfig = integrations.github
+    .list()
+    .find(integration => integration.config.host === host)?.config;
+
+  if (!integrationConfig) {
+    throw new InputError(`No integration for host ${host}`);
+  }
+
+  const credentialsProvider = GithubCredentialsProvider.create(
+    integrationConfig,
+  );
+
+  if (!credentialsProvider) {
+    throw new InputError(
+      `No matching credentials for host ${host}, please check your integrations config`,
+    );
+  }
+
+  const { token } = await credentialsProvider.getCredentials({
+    url: `${host}/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}`,
+  });
+
+  if (!token) {
+    throw new InputError(
+      `No token available for host: ${host}, with owner ${owner}, and repo ${repo}`,
+    );
+  }
+
+  return new Octokit({
+    auth: token,
+    baseUrl: integrationConfig.apiBaseUrl,
+  });
+};
+export const defaultClientFactory = async (
+  integrations: ScmIntegrationRegistry,
+  { owner, repo, host = 'github.com' }: GithubPullRequestActionInput,
+): Promise<PullRequestCreator> => {
+  const baseClient = await createClientFactory(integrations, {
+    owner,
+    repo,
+    host,
+  });
+  const client = createPullRequest(baseClient);
```

Comment:
wonder if we can return the `Octokit.plugin(createPullRequest)` here? Or does that mess up the types?

Reply:
It's more that I think that mutates Octokit globally which is _probably_ fine but maybe isn't?

Reply:
Does it? Maybe I misread the docs. 

But looks like it's non mutating:

```ts
const MyOctokit = Octokit.plugin(createPullRequest);

const TOKEN = "secret123"; // create token at https://github.com/settings/tokens/new?scopes=repo
const octokit = new MyOctokit({
  auth: TOKEN,
});
```

Reply:
D'oh, I misremembered. But it does, after all that, seem to mess with the types. Trying to work out why.

Reply:
Done, PEBKAC as usual